include/pmacc/Environment.hpp

```diff
@@ -449,26 +449,35 @@ namespace detail
         {
             throw std::runtime_error("no CUDA capable devices detected");
         }
-        else if (deviceNumber >= num_gpus) //check if device can be selected by deviceNumber
-        {
-            std::cerr << "no CUDA device " << deviceNumber << ", only " << num_gpus << " devices found" << std::endl;
-            throw std::runtime_error("CUDA capable devices can't be selected");
-        }
 #endif
 
         int maxTries = num_gpus;
-#if (PMACC_CUDA_ENABLED == 1)
-        cudaDeviceProp devProp;
-        CUDA_CHECK((cuplaError_t)cudaGetDeviceProperties(&devProp, deviceNumber));
-        /* if the gpu compute mode is set to default we use the given `deviceNumber` */
-        if (devProp.computeMode == cudaComputeModeDefault)
-            maxTries = 1;
-#endif
+        bool deviceSelectionSuccessfull = false;
+
         cudaError rc;
 
+        // search the first selectable device in the compute node
         for (int deviceOffset = 0; deviceOffset < maxTries; ++deviceOffset)
         {
             const int tryDeviceId = (deviceOffset + deviceNumber) % num_gpus;
+
+            log<ggLog::CUDA_RT>("Trying to allocate device %1%.") % tryDeviceId;
+#if (PMACC_CUDA_ENABLED == 1)
+            cudaDeviceProp devProp;
+            CUDA_CHECK((cuplaError_t)cudaGetDeviceProperties(&devProp, tryDeviceId));
+
+            /* if the cuda gpu compute mode is set to 'default'
+             * (https://docs.nvidia.com/cuda/cuda-c-programming-guide/#compute-modes)
+             * than we will use the first device. The device id selected based on
```

Comment:
I do not understand what the last sentence means. Can you please clarify *how* this depends on the local MPI rank? (And we should fix the grammar in the sentence.)

I also wonder if this routine is potentially prone to race conditions during the selection?

Reply:
In the case we are working on a process exclusive GPU system there are no races. On a system where the gpu is running in default mode, we use as gpu index the local MPI rank. This means there is no issue with races.